openedx/openedx-event-sink-clickhouse
# feat: allow to extend event sink

**Description:** This PR adds:

- `ModelBaseSink` class to easily create event sinks.
- `EVENT_SINK_CLICKHOUSE_MODEL_CONFIG` setting so that if a model changes its package location, the path can be modified from the settings without having to update this package.

**ISSUE:** 

**Dependencies:** None

**Merge deadline:** None

**Installation instructions:** List any non-trivial installation
instructions.

**Testing instructions:**

After running a clickhouse installation (you can install aspects for this), update your settings with your clickhouse credentials.
- Create a table for the user_profile_sink. You can use the patch `clickhouse-extra-sql` for that:
```yaml
name: settings
version: 0.1.0
patches:
  clickhouse-extra-sql: |
    CREATE TABLE IF NOT EXISTS {{ ASPECTS_EVENT_SINK_DATABASE }}.user_profile
    (
        id Int32 NOT NULL,
        user_id Int32 NOT NULL,
        name String NOT NULL,
        meta String NOT NULL,
        courseware String NOT NULL,
        language String NOT NULL,
        location String NOT NULL,
        year_of_birth String NOT NULL,
        gender String NOT NULL,
        level_of_education String NOT NULL,
        mailing_address String NOT NULL,
        city String NOT NULL,
        country String NOT NULL,
        state String NOT NULL,
        goals String NOT NULL,
        bio String NOT NULL,
        profile_image_uploaded_at String NOT NULL,
        phone_number String NOT NULL,
        dump_id UUID NOT NULL,
        time_last_dumped String NOT NULL
    ) engine = MergeTree
        PRIMARY KEY (id, time_last_dumped)
        ORDER BY (id, time_last_dumped);    
```
- Run `tutor dev do init -l aspects` to execute this SQL.
- Go to your account page http://apps.local.overhang.io:1997/account and update your name.
- In clickhouse, there should be a record for your update.

There is also an implementation for a generic backfill command:

- Backfill user_profile objects with ID 5
```sh
python manage.py cms dump_data_to_clickhouse --object user_profile --ids 5
```

- Backfill all user_profile objects
```sh
python manage.py cms dump_data_to_clickhouse --object user_profile
```

**Reviewers:**

**Merge checklist:**

**Post merge:**
      finished.

**Author concerns:** List any concerns about this PR - inelegant
solutions, hacks, quick-and-dirty implementations, concerns about
migrations, etc.
